core/chains/evm/log/registrations.go

```diff
@@ -183,77 +196,79 @@ func filtersContainValues(topicValues []common.Hash, filters [][]Topic) bool {
 	return true
 }
 
-func newSubscribers(evmChainID big.Int) *subscribers {
-	return &subscribers{
-		handlers:   make(map[common.Address]map[common.Hash]map[Listener]*listenerMetadata),
-		evmChainID: evmChainID,
+func newHandlers(evmChainID big.Int) *handlers {
+	return &handlers{
+		handlersByAddr: make(map[common.Address]map[common.Hash]map[Listener]*listenerMetadata),
+		evmChainID:     evmChainID,
 	}
 }
 
-func (r *subscribers) addSubscriber(reg registration) (needsResubscribe bool) {
-	addr := reg.opts.Contract
+func (r *handlers) addSubscriber(sub *subscriber) (needsResubscribe bool) {
+	addr := sub.opts.Contract
 
-	if reg.opts.MinIncomingConfirmations <= 0 {
-		reg.opts.MinIncomingConfirmations = 1
+	if sub.opts.MinIncomingConfirmations <= 0 {
+		sub.opts.MinIncomingConfirmations = 1
 	}
 
-	if _, exists := r.handlers[addr]; !exists {
-		r.handlers[addr] = make(map[common.Hash]map[Listener]*listenerMetadata)
+	if _, exists := r.handlersByAddr[addr]; !exists {
+		r.handlersByAddr[addr] = make(map[common.Hash]map[Listener]*listenerMetadata)
 	}
 
-	for topic, topicValueFilters := range reg.opts.LogsWithTopics {
-		if _, exists := r.handlers[addr][topic]; !exists {
-			r.handlers[addr][topic] = make(map[Listener]*listenerMetadata)
+	for topic, topicValueFilters := range sub.opts.LogsWithTopics {
+		if _, exists := r.handlersByAddr[addr][topic]; !exists {
+			r.handlersByAddr[addr][topic] = make(map[Listener]*listenerMetadata)
 			needsResubscribe = true
 		}
 
-		r.handlers[addr][topic][reg.listener] = &listenerMetadata{
-			opts:    reg.opts,
+		r.handlersByAddr[addr][topic][sub.listener] = &listenerMetadata{
+			opts:    sub.opts,
 			filters: topicValueFilters,
 		}
 	}
 	return
 }
 
-func (r *subscribers) removeSubscriber(reg registration) (needsResubscribe bool) {
-	addr := reg.opts.Contract
+func (r *handlers) removeSubscriber(sub *subscriber) (needsResubscribe bool) {
+	addr := sub.opts.Contract
 
-	if _, exists := r.handlers[addr]; !exists {
+	// FIXME: What about the case where you remove/add a job with the same contract address?
+	// addr is not good enough to be a unique key
```

Comment:
I suspect this might be causing the bug where jobs can be removed/re-added but don't resubscribe

Reply:
I will fix in a followup PR